helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -162,12 +162,19 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
           int statePriority1 = replica1.getStatePriority();
           int statePriority2 = replica2.getStatePriority();
           if (statePriority1 == statePriority2) {
-            // If state prioritizes are the same, compare the names.
-            if (resourceName1.equals(resourceName2)) {
-              return replica1.getPartitionName().compareTo(replica2.getPartitionName());
-            } else {
-              return resourceName1.compareTo(resourceName2);
-            }
+            // If state prioritizes are the same, try to randomize the replicas order. Otherwise,
+            // the same replicas might always be moved in the rebalance. This is because their
+            // rebalance calculating will always happen at the critical moment while the cluster is
+            // almost full.
```

Comment:
Why is the same order a problem? Do we not want it to be deterministic?

Reply:
The problem of fixed order is "the same replicas might always be moved in the rebalance", as mentioned in the comments.

If you read the following sentence in the comments, the order is still deterministic given the input is the same. There is no randomness.

Reply:
What's confusing me is that usually we want a static order in rebalance so we know what to expect, and that way it's easier for us to debug, track, and test. That has been the case with CrushED rebalancer where we sort all resources and partitions by lexical order.

So the fact that you're avoiding the static order in which replicas are given to the rebalancer is counter-intuitive. So I'd like to understand why it's different in the WAGED rebalancer setting. Also, the description claims that the static ordering makes the rebalance unstable. How so?

Reply:
I see.
The thing is that we still need a stable order given the TOPOLOGY is the same. But, we don't want a fixed order for the WAGED rebalancer.
Because of the greedy algorithm, there always be an unstable window while we try to place the replica one by one. For example, if you have 10 replicas to be assigned, usually, the rebalancer always tries to move replica 9,10. If we have a fixed order, the same replica will be moved again and again. And due to the new placement tend to be unstable, this double the possibility that the replicas being moved. So as a result, these replicas will almost be certain to be moved again in the next round of rebalance. That, as I noted in the comments, makes the assignment unstable.
The details are hard to note down in the comments. I plan to explain later in some blogs in detail.